packages/expo-cli/src/commands/upload/submission-service/archive-source/ArchiveFileSource.ts

```diff
@@ -0,0 +1,295 @@
+}
+
+interface ArchiveFileUrlSource extends ArchiveFileSourceBase {
+  sourceType: ArchiveFileSourceType.url;
+  url: string;
+}
+
+interface ArchiveFileLatestSource extends ArchiveFileSourceBase {
+  sourceType: ArchiveFileSourceType.latest;
+}
+
+interface ArchiveFilePathSource extends ArchiveFileSourceBase {
+  sourceType: ArchiveFileSourceType.path;
+  path: string;
+}
+
+interface ArchiveFileBuildIdSource extends ArchiveFileSourceBase {
+  sourceType: ArchiveFileSourceType.buildId;
+  id: string;
+}
+
+interface ArchiveFilePromptSource extends ArchiveFileSourceBase {
+  sourceType: ArchiveFileSourceType.prompt;
+  projectDir: string;
+}
+
+export type ArchiveFileSource =
+  | ArchiveFileUrlSource
+  | ArchiveFileLatestSource
+  | ArchiveFilePathSource
+  | ArchiveFileBuildIdSource
+  | ArchiveFilePromptSource;
+
+async function getArchiveFileLocationAsync(
+  mode: SubmissionMode,
+  source: ArchiveFileSource
+): Promise<string> {
+  switch (source.sourceType) {
+    case ArchiveFileSourceType.prompt:
+      return await handlePromptSourceAsync(mode, source);
+    case ArchiveFileSourceType.url: {
+      const url = await handleUrlSourceAsync(mode, source);
+      return await getArchiveLocationForUrlAsync(mode, url);
+    }
+    case ArchiveFileSourceType.latest: {
+      const url = await handleLatestSourceAsync(mode, source);
+      return await getArchiveLocationForUrlAsync(mode, url);
+    }
+    case ArchiveFileSourceType.path: {
+      const path = await handlePathSourceAsync(mode, source);
+      return getArchiveLocationForPathAsync(mode, path);
+    }
+    case ArchiveFileSourceType.buildId: {
+      const url = await handleBuildIdSourceAsync(mode, source);
+      return await getArchiveLocationForUrlAsync(mode, url);
+    }
+  }
+}
+
+async function getArchiveLocationForUrlAsync(mode: SubmissionMode, url: string): Promise<string> {
+  if (mode === SubmissionMode.online) {
+    return url;
+  } else {
+    const tmpPath = pathJoin(os.tmpdir(), pathBasename(url));
+    log('Downloading your app archive');
+    return downloadAppArchiveAsync(url, tmpPath);
+  }
+}
+
+async function getArchiveLocationForPathAsync(mode: SubmissionMode, path: string): Promise<string> {
+  if (mode === SubmissionMode.online) {
+    log('Uploading your app archive to the temporary storage');
+    return await uploadAppArchiveAsync(path);
+  } else {
+    return path;
+  }
+}
+
+async function handleUrlSourceAsync(
+  mode: SubmissionMode,
+  source: ArchiveFileUrlSource
+): Promise<string> {
+  return source.url;
+}
+
+async function handleLatestSourceAsync(
+  mode: SubmissionMode,
+  source: ArchiveFileLatestSource
+): Promise<string> {
+  const { owner, slug } = getAppConfig(source.projectDir);
+  const builds = await StandaloneBuild.getStandaloneBuilds(
+    {
+      platform: source.platform,
+      owner,
+      slug,
+    },
+    1
+  );
+  if (builds.length === 0) {
+    log.error("Couldn't find any builds for this project.");
```

Comment:
we should be more clear about what this means and follow guidelines as mentioned in https://github.com/expo/expo-cli/pull/2165

Reply:
Actually, this command won't exit if this log is printed. See the line below - a prompt will appear. 
However, I'll **try** to make this message a bit more clear.